### PR TITLE
Pyffi overhaul

### PR DESCRIPTION
**Pull request by Candoran2**
@niftools/blender-niftools-addon-reviewer 

# Overview
1. Changes kf and nif import/export to make use of the new nif statically generated nif library.
2. Added support for SSE meshes (import only).

##  Detailed Description
1. Since most functions were ported straight from pyffi, functionally little code needed to be changed. The classes are still easily accessible, but under `generated.formats.nif` in the `classes` variable (it's actually a dict that also allows value access through `dict.key` in addition to `dict['key']`).
2. SSE nifs use BSTriShapes, so the import process had to be partially adjusted. Getting the geometry information from the nif (vertex positions, triangles etc) was decoupled from setting it. BSTriShapes were given the `skin_partition` property as alternative access to the `skin` field because it pretty much refers to the same thing as the `skin_partition` field on NiGeometry. Ideally this would be adjusted in the xml so it isn't necessary, but that would probably be a breaking change for other software that use the xml.

## Fixes Known Issues
- Doesn't completely fix, but is a first step towards #473, #474, #480, #534 and I'm sure a number of others.
- #521 

## Documentation
**[Overview of updates to documentation]**

## Testing
**[Overview of testing required to ensure functionality is correctly implemented]**

### Manual
**[Set of steps to manually verify updates are working correctly]**

### Automated
**[List of tests run, updated or added to avoid future regressions]**

## Additional Information
**[Anything else you deem relevant]**

### CODE REVIEW

**Comment by HENDRIX-ZT2** [comment]
Could we change the import to put the .classes part into the import statements instead of each NifFormat call? Not sure if that works though

**Comment by Candoran2** [comment]
> Could we change the import to put the .classes part into the import statements instead of each NifFormat call? Not sure if that works though

Yes. I just did it.

File: io_scene_niftools/modules/nif_import/object/__init__.py
Comment (HENDRIX-ZT2) [inline io_scene_niftools/modules/nif_import/object/__init__.py:148]: Enum access?

File: io_scene_niftools/modules/nif_import/property/geometry/mesh.py
Comment (HENDRIX-ZT2) [inline io_scene_niftools/modules/nif_import/property/geometry/mesh.py:73]: put both in one list?

File: io_scene_niftools/modules/nif_import/property/shader/bsshaderproperty.py
Comment (HENDRIX-ZT2) [inline io_scene_niftools/modules/nif_import/property/shader/bsshaderproperty.py:93]: enum access?

**Review by HENDRIX-ZT2 — CHANGES_REQUESTED** [review]
A few remarks, one important todo. Stellar job so far!

File: io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py
Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py:427]: Yes. `from_value` (originally the __new__ method as generated by the standard lib's EnumMeta class) does conversion from ints to enum members. Blender stores the consistency types as strings, so you need to do getattr to simulate ConsistencyType.MUTABLE, for example.

File: io_scene_niftools/modules/nif_export/property/shader/__init__.py
Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/property/shader/__init__.py:96]: See previous response.

File: io_scene_niftools/modules/nif_import/collision/__init__.py
Comment (Candoran2) [reply io_scene_niftools/modules/nif_import/collision/__init__.py:107]: Just testing it out, they seem to have both. The difference being that name is read-only, whereas \_name\_ is not, so we should probably use the former.

File: io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py
Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py:427]: In that case, can't we do NifClasses.ConsistencyType[b_obj.niftools.consistency_flags]?
https://docs.python.org/3/library/enum.html#programmatic-access-to-enumeration-members-and-their-attributes

Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/geometry/mesh/__init__.py:427]: Yes, you're right.

File: io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py
Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py:147]: Not quite. In BSLightingShaderProperty (and other places), it's a proper enum with int storage. In BSEffectShaderProperty, it's a byte, but with the same meaning (next to three other byte-wide fields). There's a couple ways to resolve that in the xml if you want it to be a proper enum member in the BSEffectShaderProperty:

- Create a bitfield in the BSEffectShaderProperty with an enum member for the texture clam mode as a member of the bitfield, leave the BSLightingShaderProperty (and other places) as-is. Downside is that then access in the BSEffectShaderProperty is different from access of the same field in the BSLightingShaderProperty .
- Create a bitfield in the BSEffectShaderProperty _and_ the BSLightingShaderProperty (and other places). That does mean you create two members for the BSLightingShaderProperty that don't actually occur there (as far as I know, haven't actually tested whether those bytes do nothing), but at least access is the same.
- Change the storage of the TexClampMode enum, and add unused bytes in the BSLightingShaderProperty (and other places).

However, as long as you do `==` comparison it doesn't actually matter whether it's an int or a proper enum member. Due to the way the xml is at the moment, the `% 256` has become unnecessary, and it's functional for now.

Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py]: > * Change the storage of the TexClampMode enum, and add unused bytes in the BSLightingShaderProperty (and other places).
I think this is the proper way. Use a ubyte storage, and pad where needed.

Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py]: Actually, come to think of it, that will probably not work correctly for different endiannesses - due to that, there is a functional difference between a byte enum followed by three unused bytes and a uint of which the least significant byte is used.

That leaves the first and second option, or leaving it as-is.

Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py]: You are right, I forget that nif has variable endianness.

Comment (Candoran2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py]: I think it's fine to leave as-is in the addon for now. I've made an issue on the xml github for future followup.

Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_export/property/texture/types/bsshadertexture.py]: Right!

File: io_scene_niftools/modules/nif_import/property/texture/loader.py
Comment (Candoran2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py:142]: Do you have a nif with which the original function didn't work and the new one does, so I can test it?

File: io_scene_niftools/utils/blocks.py
Comment (Candoran2) [reply io_scene_niftools/utils/blocks.py:4]: Is there any rhyme or reason as to which nif versions use which encodings, or is it really just trying out every encoding and seeing which works?

Comment (HENDRIX-ZT2) [reply io_scene_niftools/utils/blocks.py:4]: This is an interesting thought. But generally, it depends on the specific game (studio). I think the try workflow is good enough for now.

File: io_scene_niftools/modules/nif_import/property/texture/loader.py
Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py:142]: I didn't do that change, but afaik it's really just an issue in the pyffi code for internal textures.

Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py:142]: Here's one sample nif with internal textures that I believe we we're testing stuff on:
[AS50_SNAKE.zip](https://github.com/niftools/blender_niftools_addon/files/9741738/AS50_SNAKE.zip)

Comment (Candoran2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py:141]: I don't think that one has internal textures as far as I can tell, or at least not the blocks that are affected by this code (ATextureRenderData/NiPixelFormat and the two blocks that inherit from it).

Comment (HENDRIX-ZT2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py:141]: My bad, this one does:
[TDM_PanicSpace_c.zip](https://github.com/niftools/blender_niftools_addon/files/9745378/TDM_PanicSpace_c.zip)

Comment (Candoran2) [reply io_scene_niftools/modules/nif_import/property/texture/loader.py]: Updated. Had to make one small change to the DdsFile class of cobra-tools, though.

**Review by HENDRIX-ZT2 — APPROVED** [review]

**Comment by Allofich** [comment]
Hello. I've been keeping an eye on this project as I'm interested in it. May I ask if there is an update on this PR?

I was motivated to comment because, when looking at open PRs at the PyFFI repository, I saw https://github.com/niftools/pyffi/pull/74, which looks, at least, like a similar situation to this PR. It has one reviewer giving it an OK, with another requested to review, and seems to have remained in that state for 2 years. I feel it would be a shame for this PR to end up in a similar situation. The contributor might leave when they could have instead been motivated to continue sending in additional improvements if the PR was accepted.

So, this is just to let you know that those like myself who are interested in the project, and either a merge of PRs like this or the one I linked to, or a comment about why merging is not yet possible, etc. would be appreciated and keep interest going. Thanks.

**Comment by Candoran2** [comment]
> Hello. I've been keeping an eye on this project as I'm interested in it. May I ask if there is an update on this PR?
> 
> I was motivated to comment because, when looking at open PRs at the PyFFI repository, I saw [niftools/pyffi#74](https://github.com/niftools/pyffi/pull/74), which looks, at least, like a similar situation to this PR. It has one reviewer giving it an OK, with another requested to review, and seems to have remained in that state for 2 years. I feel it would be a shame for this PR to end up in a similar situation. The contributor might leave when they could have instead been motivated to continue sending in additional improvements if the PR was accepted.
> 
> So, this is just to let you know that those like myself who are interested in the project, and either a merge of PRs like this or the one I linked to, or a comment about why merging is not yet possible, etc. would be appreciated and keep interest going. Thanks.
> 
> Edit: Looks like the contributor is a member of Niftools, so what I said about them possibly leaving may not be applicable.

Seems like Neomonkeus (who normally does the merging/releases) isn't available at the moment. My plan so far is to wait until the 19th and merge it myself if there hasn't been any update by then.

In the meantime I'm still working on the addon, In just not yet pushing the commits to GitHub.

**Comment by Allofich** [comment]
Thanks for the update. I didn't know you were able to merge it yourself (I didn't realize you were part of the Niftools team either until I looked at your profile after making my comment), and it's good to hear that work is continuing. It'll be great if we can start importing and exporting .nif files with current versions of Blender, which doesn't seem to work yet (I've only tested with Oblivion .nif files, so I don't know if other games work).

**Comment by Candoran2** [comment]
> Thanks for the update. I didn't know you were able to merge it yourself (I didn't realize you were part of the Niftools team either until I looked at your profile after making my comment), and it's good to hear that work is continuing. It'll be great if we can start importing and exporting .nif files with current versions of Blender, which doesn't seem to work yet (I've only tested with Oblivion .nif files, so I don't know if other games work).

Importing and exporting nif files with current versions of Blender should already be possible with the released versions of the plugin (especially Oblivion nifs, which have long been supported) - if you're encountering errors while doing that, please make an issue and post the concerning nifs so that we can help.

**Comment by Allofich** [comment]
The problems I encountered are covered by https://github.com/niftools/blender_niftools_addon/issues/453. Trying to import meshes/architecture/anvil/anvilaltar01.nif failed because of "'int' object has no attribute 'material'" errors. A fix is mentioned in that issue report, but I don't know if it's a correct fix.

The exported .nif then supposedly lacks collision in-game. I can't remember if I tested that part myself, but I was left with the impression that the plugin wasn't functional yet, at least for Oblivion .nif files.

**Comment by Allofich** [comment]
Well I just downloaded the latest Blender release (3.3), installed the latest blender_niftools_addon, and meshes/architecture/anvil/anvilaltar01.nif imports without an error now. I thought I had already tested the latest blender_niftools_addon and found that it still didn't work, and I thought that no changes related to the "material" attribute had been made, so it's a bit of a surprise. https://github.com/niftools/blender_niftools_addon/issues/453 may need updating, then. I'm going to keep playing around with it for a while.

**Comment by Allofich** [comment]
I remembered I had this conversation https://github.com/niftools/pyffi/pull/69 over at the Pyffi project, where I was told the "material" workaround wasn't a good solution (it would "break other things"). I was told you and HENDRIX-ZT2 were working on a proper solution. So that's where I got the impression things were WIP and it would be some time before Oblivion .nifs could be exported successfully.

I don't know how common the "material" issue is with Oblivion .nifs, but testing a few others I quickly came upon another with the same error: meshes/architecture/anvil/anvilfgfirstfloor.nif.

Anyway I didn't mean to get off-topic. I know this PR is not about the issues I am mentioning.

**Comment by HENDRIX-ZT2** [comment]
I think we can just go ahead and merge this. IMO we don't need to do micro-releases after every PR. It just clogs down development.

**Comment by enpinion** [comment]
What is the 'generated' module actually? Is it missing directory to be pushed?
Because if I pull the current develop branch and build blender throws generated modules are missing. It looks similar as pyffi but that module has no array::Array.

**Comment by enpinion** [comment]
Thank you for the reply. Now it works.

**Comment by Allofich** [comment]
@Candoran2 Could you explain a bit more how to create the needed module?

I've downloaded the latest cobra-tools and followed the instructions to install Python and run "pip install pyqt5 imageio". I've also downloaded the latest sources from https://github.com/Candoran2/new-pyffi and https://github.com/niftools/nifxml/tree/develop (which seems to be newer than https://github.com/Candoran2/nifxml/tree/develop), but I don't know how to proceed and can't find any information on using codegen with pyffi.

**Comment by Candoran2** [comment]
> @Candoran2 Could you explain a bit more how to create the needed module?
> 
> I've downloaded the latest cobra-tools and followed the instructions to install Python and run "pip install pyqt5 imageio". I've also downloaded the latest sources from https://github.com/Candoran2/new-pyffi and https://github.com/niftools/nifxml/tree/develop (which seems to be newer than https://github.com/Candoran2/nifxml/tree/develop), but I don't know how to proceed and can't find any information on using codegen with pyffi.

For simple operation: Copy or softlink the `formats/nif` folder from pyffi into the cobra-tools `source/formats` folder. Make sure that the nif xml is also in that source/formats/nif folder. Make sure the `spells` and the `utils` folder from new-pyffi are similarly in the `source` folder of cobra-tools.

Then, simply run codegen.py. There's a lot of other code in the cobra-tools repository, but that and the codegen folder are the only part that we need (and some other base format libraries like base and dds that get generated similarly). This will fill the `generated` folder in cobra-tools with a lot of python code, including that necessary for reading nifs. You can copy or link that in the dependencies folder of the niftools addon, and everything should work.

**Comment by Allofich** [comment]
Thank you for the instructions. I was able to get the plugin working (with problems, see below) by running `codegen.py` and then copying the `generated` folder into the `dependencies` folder of the niftools addon (and then zipping the addon into a .zip file for installation in Blender).

With only the `generated` folder in `dependencies`, the options for importing or exporting .NIF, etc. were not added to the import menu, so I tried adding the dependencies as taken from the latest release (v0.014): the `bin`, `pyffi` and `PyFFI-2.2.4.dev3.dist-info` folders. With these added to the `dependencies` folder, the import and export options appeared, but are these dependencies from the v0.014 release file suitable for the newer code? I'm trying the latest `develop` branch code (f5211a9). I haven't tried the newer code from https://github.com/niftools/blender_niftools_addon/pull/541.

While the addon is at least working to some degree, I'm seeing errors and actually get somewhat less far in my test case (`meshes/architecture/anvil/anvilaltar01.nif` from Oblivion) than with the v0.014 release. I don't know if this is just because it is WIP, or if I maybe have an incompatibility in the dependencies.

I'm testing two versions of anvilaltar01.nif. One is the version you get from installing from GOG or Steam. The other is from the DVD release (the DVD release has slightly different mesh files that report a later NIF version).

With the v0.014 addon both versions of anvilaltar01.nif import successfully. They both fail to export due to the problems mentioned in https://github.com/niftools/blender_niftools_addon/issues/453.

With the newer addon code from the develop branch I'm trying now, only the GOG/Steam version imports successfully. The DVD version fails with
```
niftools:INFO:Executing - Niftools : Blender Niftools Addon v0.0.15(running on Blender 3.3.1, Nif xml version 0.9.3.0)
niftools:INFO:Importing ****\anvilaltar01 CD.nif
niftools:INFO:NIF file version: 14000005
niftools:INFO:Reading .nif file
Traceback (most recent call last):
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\operators\nif_import_op.py", line 134, in execute
    return NifImport(self, context).execute()
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 74, in execute
    self.load_files()  # needs to be first to provide version info.
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 142, in load_files
    scene.import_version_info(NifData.data)
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\modules\nif_import\scene\__init__.py", line 55, in import_version_info
    scene.game = possible_games[0].name
TypeError: bpy_struct: item.attr = val: enum "UNKNOWN" not found in ('NONE', 'ATLANTICA_ONLINE', 'AURA_KINGDOM', 'AXIS_AND_ALLIES', 'BLOOD_BOWL', 'BULLY_SE', 'CIVILIZATION_IV', 'CULPA_INNATA', 'DARK_AGE_OF_CAMELOT', 'DIGIMON_MASTERS_ONLINE', 'DIVINITY_2', 'DIVINITY_2_0_X_10000', 'EMERGE', 'EMPIRE_EARTH_II', 'EMPIRE_EARTH_III', 'FALLOUT_3', 'FALLOUT_4', 'FALLOUT_76', 'FALLOUT_NV', 'FANTASY_FRONTIER', 'FFT_ONLINE', 'FLORENSIA', 'FREEDOM_FORCE', 'FREEDOM_FORCE_VS_THE_3_RD_REICH', 'GHOST_IN_THE_SHELL_FIRST_ASSAULT', 'HOWLING_SWORD', 'IRIS_ONLINE', 'KOHAN_2', 'KRAZY_RAIN', 'LAZESKA', 'LEGO_UNIVERSE', 'LOKI', 'MAPLE_STORY_2', 'MICRO_VOLTS', 'MORROWIND', 'MUNCH_S_ODDYSEE', 'NEO_STEAM', 'OBLIVION', 'OBLIVION_KF', 'PRISON_TYCOON', 'PRO_CYCLING_MANAGER', 'RAGNAROK_ONLINE_2', 'RED_OCEAN', 'ROCKSMITH', 'ROCKSMITH_2014', 'SHIN_MEGAMI_TENSEI_IMAGINE', 'SID_MEIER_S_PIRATES', 'SID_MEIER_S_RAILROADS', 'SKYRIM', 'SKYRIM_SE', 'THE_GUILD_2', 'WARHAMMER', 'WILDLIFE_PARK_2', 'WIZARD_101', 'WORLD_SHIFT', 'ZOO_TYCOON_2')
```

The GOG/Steam version successfully imports but has an error
```
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\operators\nif_import_op.py", line 134, in execute
    return NifImport(self, context).execute()
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 131, in execute
    self.import_root(root)
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 159, in import_root
    b_obj = self.import_branch(root_block)
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 240, in import_branch
    b_children.extend(self.import_collision(n_block))
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\nif_import.py", line 184, in import_collision
    return self.bhkhelper.import_bhk_shape(n_node.collision_object.body)
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\modules\nif_import\collision\havok.py", line 93, in import_bhk_shape
    return self.process_bhk(bhk_shape)
  File "****\Blender 3.3\3.3\python\lib\functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\modules\nif_import\collision\havok.py", line 151, in import_bhk_ridgidbody_t
    b_rot = bhk_shape.rotation
AttributeError: 'bhkRigidBodyT' object has no attribute 'rotation'
```

When trying to export, it fails with the following
```
niftools:INFO:Exporting
niftools:INFO:Exporting geometry and animation
niftools:INFO:Writing NIF version 0x14000004
niftools:INFO:Created meta root because blender scene had 11 root objects
niftools:INFO:Exporting NiNode block
niftools:INFO:Exporting <bpy_struct, Object("AnvilAltar01") at 0x000001DC552F5D08> as NiNode block
niftools:INFO:Exporting <bpy_struct, Object("AltarCloth01") at 0x000001DC552F6408> as NiNode block
niftools:INFO:Exporting <bpy_struct, Object("AltarCloth01:0") at 0x000001DC552F4F08>
niftools:INFO:Exporting <bpy_struct, Object("AltarCloth01:0") at 0x000001DC552F4F08> as NiTriShape block
niftools:INFO:Exporting <bpy_struct, Object("AltarCloth01:0") at 0x000001DC552F4F08> as NiTriShapeData block
niftools:INFO:Exporting NiStencilProperty block
niftools:INFO:Exporting NiSpecularProperty block
niftools:INFO:Exporting NiMaterialProperty block
niftools:ERROR:Export version expected shader. No shader applied to mesh '<bpy_struct, Material("RFINTxIron") at 0x000001DC52D61008, evaluated>', these cannot be exported to NIF.Set shader before exporting.
niftools:ERROR:****\AppData\Roaming\Blender Foundation\Blender\3.3\scripts\addons\io_scene_niftools\modules\nif_export\property\shader\__init__.py:56
```

Any ideas? Should I try https://github.com/niftools/blender_niftools_addon/pull/541? Could the dependencies I used be incompatible?

**Comment by Candoran2** [comment]
So the generated module is indeed in addition to the existing dependencies, you can't remove the old ones quite yet. The first error is indeed fixed in #541, not sure about the second one. It very well might be fixed there (I did make some updates for collision) but it's possible I missed some places.

As for export, I haven't made any huge changes there. Just note that, since Oblivion has a lot of nif versions, it's probably safest to set the game yourself right before exporting.

**Comment by Allofich** [comment]
I updated to Blender 3.4 (I doubt that makes a difference) and am testing the newer code from https://github.com/niftools/blender_niftools_addon/pull/541 now.

I can confirm the DVD version of anvilaltar01.nif imports now, although it has the same `AttributeError: 'bhkRigidBodyT' object has no attribute 'rotation'` error as the GOG/Steam version.

After import, the DVD version of anvilaltar01.nif has game set to `UNKNOWN`. After changing this to `Oblivion`, it successfully exported, with some warnings.

```
niftools:WARNING:Mopps for non-static objects may not function correctly in-game. You may wish to use simple primitives for collision.
niftools:INFO:Writing .nif file
WARNING:root:instance 0 is not a member of the <enum 'OblivionLayer'> class
WARNING:root:instance 1 is not a member of the <enum 'hkResponseType'> class
WARNING:root:instance 0 is not a member of the <enum 'OblivionLayer'> class
WARNING:root:instance 1 is not a member of the <enum 'hkResponseType'> class
WARNING:root:instance 1 is not a member of the <enum 'SpecularFlags'> class
```

The GOG/Steam .nif also imports (with the above-mentioned error), but the game is automatically set to `Fallout 3`. (I didn't notice if this happened with the develop branch code I tested earlier) I manually changed it to `Oblivion` and it exported successfully, with the same warnings as the DVD .nif.

Now time to go check the exported .nifs in-game.

**Comment by Allofich** [comment]
Both the exported .nifs crash Oblivion when I try to load a saved game where I'm standing in front of the altar model.

**Comment by Candoran2** [comment]
> Both of the exported .nifs crash Oblivion when I try to load a saved game where I'm standing in front of the altar model. (I tried replacing the .nif with another .nif that comes with the game and that did not crash, so I think the crash is being caused by the exported .nif files)
> 
> Edit: Just to be sure I tried starting a new game (rather than loading a saved game) and warping to the cell with the .nif with the console, and the crash still happened.

Probably best to start an issue at this point.

**Comment by Allofich** [comment]
Alright, I'll do so. Thanks again for helping me get things up and running so I could test the new code.

**Comment by Allofich** [comment]
Looks like you've made some fixes, so I'll re-build and re-test before making any issues.

**Pull request merged**
